docs/architecture/deployment-execution/state.md

```diff
@@ -33,6 +33,33 @@ remote HTTP service, which is responsible for managing the underlying state.
 [Pulumi Cloud](https://www.pulumi.com/product/pulumi-cloud/) is the primary
 example of this.
 
+(snapshots)=
+## Snapshots
+
+The Pulumi CLI generates a new snapshot at the start and at the end of each
+operation, unless the `PULUMI_SKIP_CHECKPOINTS` environment variable is set.
+
+Currently we upload either a full snapshot, or the diff to the previous snapshot
+if the HTTP state backend is used, and the snapshot is sufficiently large.
+
+Depending on the step type, at the start of the operation we create a "pending
+operation" entry. This is created, so if anything goes wrong during the operation,
+e.g. the connection drops while we set up a resource, we have a record of it, and
+the user can manually go through them and either delete the entry if the resource
+has not been created in the provider, or import the resource into the state. At
+the end of each operation, we finalize the entry, and add it to the list of
+resouces, while removing the "pending operation" entry.
+
+Note that the engine is currently free to modify the snapshot in any way, and we'll
+always upload that internal snapshot, since the snapshot manager internally uses
+a pointer to the same snapshot as the engine is using.
```

Comment:
The engine makes heavy use of this. Both for tagging resources as to be deleted because of replace operations (setting `Delete=true`) and for changing `Outputs` due to a `RegisterResourceOutputs` call.

Reply:
Good call, let me emphasize this a little more in the text.  I think it's also used for refreshes (that are not marked as "Persisted()"), and default provider updates. Added a note for that as well.